grails-app/services/au/org/ala/collectory/EmlImportService.groovy

```diff
@@ -137,36 +138,77 @@ class EmlImportService {
 
             eml.dataset.metadataProvider.each {
-                def contact = addContact(it)
+                def contact = addOrUpdateContact(it)
                 if (contact){
                     contacts << contact
                 }
             }
         }
 
-        contacts
+        // Add additional contacts
+        if (eml.dataset.contact){
+            eml.dataset.contact.each {
+                def contact = addOrUpdateContact(it)
+                if (contact){
+                    contacts << contact
+                    primaryContacts << contact
+                }
+            }
+        }
+
+        if (eml.dataset.associatedParty){
+            eml.dataset.associatedParty.each {
+                def contact = addOrUpdateContact(it)
+                if (contact){
+                    contacts << contact
+                }
+            }
+        }
+
+        [contacts: contacts, primaryContacts: primaryContacts]
     }
 
-    private def addContact(emlElement){
-        def contact = Contact.findByEmail(emlElement.electronicMailAddress)
-        if (!contact){
+    private def addOrUpdateContact(emlElement) {
+        def contact = null
+        if (emlElement.electronicMailAddress && !emlElement.electronicMailAddress.isEmpty()) {
+            String email = emlElement.electronicMailAddress.text().trim()
+            contact = Contact.findByEmail(email)
+        } else if (emlElement.individualName.givenName && emlElement.individualName.surName) {
+            contact = Contact.findByFirstNameAndLastName(emlElement.individualName.givenName, emlElement.individualName.surName)
+        } else if (emlElement.individualName.surName) {
+            // surName is mandatory
+            contact = Contact.findByLastName(emlElement.individualName.surName)
+        }
+
+        // Create the contact if it doesn't exist and it's a individualName with email or surName
+        // to prevent empty contacts (e.g. with emlElement.organizationName only)
+        boolean hasEmail = emlElement?.electronicMailAddress?.text()?.trim()?.isEmpty() == false
+        boolean hasName = emlElement?.individualName?.surName?.text()?.trim()?.isEmpty() == false
+
+        if (!contact && (hasEmail || hasName)) {
             contact = new Contact()
-            contact.firstName = emlElement.individualName.givenName
-            contact.lastName = emlElement.individualName.surName
-            contact.email = emlElement.electronicMailAddress
-            contact.setUserLastModified(collectoryAuthService.username())
-            Contact.withTransaction {
-                if (contact.validate()) {
-                    contact.save(flush: true, failOnError: true)
-                    return contact
-                } else {
-                    contact.errors.each {
-                        log.error("Problem creating contact: " + it)
-                    }
-                    return null
+        } else {
+            return null
```

Comment:
@adam-collins I added this to prevent the NPE if a contact does not have name or email